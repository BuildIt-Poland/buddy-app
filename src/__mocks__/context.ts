import { AuthContextData } from 'context/AuthStore';

export const authContext = (context?: Partial<AuthContextData>) => ({
  data: { role: 'BUDDY', token: 'dummy-token', userId: '1' },
  loading: false,
  error: null,
  isAuthenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  ...context,
});
