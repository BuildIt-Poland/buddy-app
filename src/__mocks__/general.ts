import { ROUTES } from 'shared/routes';

export const mockLocation = (path: string = ROUTES.BASE) => ({
  key: 'utwyk7',
  pathname: path,
});

export const mockSchemaTypes = (types: any = {}) => ({
  UserRole: {
    Newbie: 'NEWBIE',
    Buddy: 'BUDDY',
  },
  TaskStatus: {
    Completed: 'COMPLETED',
    Uncompleted: 'UNCOMPLETED',
  },
  ...types,
});
