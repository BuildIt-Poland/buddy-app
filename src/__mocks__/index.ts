export * from './general';
export * from './graphql';
export * from './schema';
