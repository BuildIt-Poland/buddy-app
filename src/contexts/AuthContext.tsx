import React from 'react';
import { AuthPayload } from 'buddy-app-schema';

export interface State {
  loading: boolean;
  error: any;
  data: AuthPayload;
  isAuthenticated: boolean;
  isAppBootstrapped: boolean;
}

export interface AuthContextData extends State {
  login: (email: string, password: string) => void;
  logout: () => void;
}

export const defaultState: State = {
  loading: false,
  error: null,
  data: {} as AuthPayload,
  isAuthenticated: false,
  isAppBootstrapped: false,
};

const defaultContext: AuthContextData = {
  ...defaultState,
  login: () => null,
  logout: () => null,
};

const AuthContext = React.createContext<AuthContextData>(defaultContext);

export default AuthContext;
