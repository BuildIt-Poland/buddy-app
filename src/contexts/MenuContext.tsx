import React from 'react';

export interface State {
  isOpen: boolean;
}

export interface MenuContextData extends State {
  toggleMenu: () => void;
}

export const defaultState: State = {
  isOpen: false,
};

const defaultContext: MenuContextData = {
  ...defaultState,
  toggleMenu: () => null,
};

const MenuContext = React.createContext<MenuContextData>(defaultContext);

export default MenuContext;
