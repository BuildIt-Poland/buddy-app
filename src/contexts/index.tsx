import React from 'react';

import { apolloClient } from 'utils';
import { ApolloProvider } from '@apollo/react-hooks';
import theme from 'styles/theme';
import ThemeProvider from '@material-ui/styles/ThemeProvider';
import AuthProvider from 'stores/AuthProvider';
import MenuProvider from 'stores/MenuProvider';
import SnackbarProvider from 'stores/SnackbarProvider';

const AppProviders: React.FC = ({ children }) => {
  return (
    <ApolloProvider client={apolloClient}>
      <ThemeProvider theme={theme}>
        <AuthProvider>
          <MenuProvider>
            <SnackbarProvider>{children}</SnackbarProvider>
          </MenuProvider>
        </AuthProvider>
      </ThemeProvider>
    </ApolloProvider>
  );
};

export default AppProviders;
