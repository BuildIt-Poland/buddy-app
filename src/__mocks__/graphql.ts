import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { BUDDY_MENU_DETAILS, NEWBIE_MENU_DETAILS } from 'graphql/user-menu.graphql';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { TASK_DETAILS } from 'graphql/task-details.graphql';
import { GraphQLError } from 'graphql';
import {
  MutationLoginArgs,
  AuthPayload,
  TaskStatus,
  Newbie,
  QueryNewbieArgs,
  QueryBuddyArgs,
  UserRole,
} from 'buddy-app-schema';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { getBasicUserDetailsMock } from './general';
import { authContext } from './context';

export const loginSuccessMock = (
  variables?: Partial<MutationLoginArgs>,
  data?: Partial<AuthPayload>
) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    data: {
      login: {
        ...authContext().data,
        ...data,
      },
    },
  },
});

export const loginFailedMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    errors: [new GraphQLError('No such user found')],
  },
});

export const loginNoNetworkMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  error: new Error('Network error'),
});

export const NewbieAvatarDetails = (variables?: QueryNewbieArgs) => ({
  request: {
    query: AVATAR_HEADER,
    variables,
  },
  result: {
    data: {
      newbie: {
        name: 'Tom Hanks',
        position: 'front-end',
        photo: 'some-url',
        tasksInfo: {
          buddyProgress: 0.54,
        },
      },
    },
  },
});

export const tasksResponse = (filter?: TaskStatus) => {
  const tasks = [
    {
      id: '1',
      title: 'Call you head manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Uncompleted,
      newbie: {} as Newbie,
    },
    {
      id: '2',
      title: 'Write you main manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Completed,
      newbie: {} as Newbie,
    },
  ];

  switch (filter) {
    case TaskStatus.Completed:
      return [tasks[0]];
    case TaskStatus.Uncompleted:
      return [tasks[1]];
    default:
      return tasks;
  }
};

export const buddyMenuDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: {
          name: 'Tom Hanks',
          email: 'tom@wipro.com',
          allowPushedNotifications: true,
          photo: 'buddy-photo-url',
          newbies: [
            {
              id: 'id',
              name: 'Sandra Bullock',
              photo: 'newbie-photo-url',
            },
          ],
        },
      },
    },
  },
];

export const newbieMenuDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: {
          name: 'Sandra Bullock',
          email: 'doejoe@wipro.com',
          allowPushedNotifications: true,
          photo: 'newbie-photo-url',
          buddy: {
            id: 'id',
            name: 'Tom Hanks',
            photo: 'newbie-photo-url',
          },
        },
      },
    },
  },
];

export const newbieContactDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: getBasicUserDetailsMock(UserRole.Newbie),
      },
    },
  },
];

export const buddyContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Buddy),
      },
    },
  },
];

export const newbieSelectMock = [
  {
    request: {
      query: NEWBIE_SELECT,
      variables: {
        buddyId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        buddy: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          newbies: [
            {
              id: 'ck17svulh9k2k0b17j31ansfk',
              photo: null,
              name: 'Test 2',
              startDate: null,
              tasksInfo: { buddyProgress: 0.5 },
            },
            {
              id: 'ck17swp2m9kcv0b17we0ibrdn',
              photo: null,
              name: 'Test 1',
              startDate: null,
              tasksInfo: { buddyProgress: 0.6 },
            },
          ],
        },
      },
    },
  },
];

export const taskDetailsMock = [
  {
    request: {
      query: TASK_DETAILS,
      variables: {
        taskId: '1',
      },
    },
    result: {
      data: {
        task: {
          title: 'New task',
          description: 'New task',
          status: TaskStatus.Uncompleted,
        },
      },
    },
  },
];
