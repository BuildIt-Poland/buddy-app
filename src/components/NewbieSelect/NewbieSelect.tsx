import React, { useContext } from 'react';
import Typography from '@material-ui/core/Typography';
import { useQuery } from '@apollo/react-hooks';
import Box from '@material-ui/core/Box';
import { QueryBuddyArgs, Query, Newbie } from 'buddy-app-schema';
import AuthContext, { AuthContextData } from 'contexts/AuthContext';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { makeStyles, Theme } from '@material-ui/core/styles';
import PlusButton from 'components/PlusButton';
import Carrousel from 'components/Carrousel';
import PageContainer from 'components/PageContainer';
import Header, { MenuTypes } from 'components/Header';
import MenuContext, { MenuContextData } from 'contexts/MenuContext';
import NewbieSelectDictionary from './dictionary';

const useStyles = makeStyles<Theme>(theme => ({
  title: {
    textAlign: 'center',
    marginBottom: theme.spacing(5),
    paddingRight: theme.spacing(7),
  },
  carrouselWarpper: {
    marginBottom: theme.spacing(2),
    paddingRight: theme.spacing(7),
  },
}));

const NewbieSelect: React.FC = () => {
  const { data: AuthData } = useContext<AuthContextData>(AuthContext);
  const { loading, data } = useQuery<Query, QueryBuddyArgs>(NEWBIE_SELECT, {
    variables: { buddyId: AuthData.userId },
  });
  const { title, carrouselWarpper } = useStyles();
  const { toggleMenu } = React.useContext<MenuContextData>(MenuContext);

  return (
    <>
      <Header
        type={'menu' as MenuTypes}
        onButtonClick={toggleMenu}
        loading={loading}
      />
      <PageContainer data-testid='newbie-select-page'>
        <Box className={title} component='section'>
          <Typography component='h1' variant='h2'>
            {NewbieSelectDictionary.TITLE}
          </Typography>
          <Typography color='textSecondary' component='p' variant='body2'>
            {NewbieSelectDictionary.SUBTITLE}
          </Typography>
        </Box>
        {data && data.buddy.newbies && (
          <Box className={carrouselWarpper} component={'section'}>
            <Carrousel newbies={data.buddy.newbies as Newbie[]} />
          </Box>
        )}
        <PlusButton disabled />
      </PageContainer>
    </>
  );
};

export default NewbieSelect;
