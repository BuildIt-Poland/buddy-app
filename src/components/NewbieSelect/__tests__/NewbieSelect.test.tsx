import React from 'react';
import { create, act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import waitForExpect from 'wait-for-expect';
import { newbieSelectMock, mockedBuddyContext } from '__mocks__';
import { AuthProvider } from 'contexts/AuthContext';
import { MenuProvider } from 'contexts/MenuContext';
import NewbieSelect from '../NewbieSelect';

jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/Box', () => 'Box');
jest.mock('components/PlusButton/', () => 'PlusButton');
jest.mock('components/NewbieGrid', () => 'NewbieGrid');
jest.mock('components/PageContainer/', () => 'PageContainer');
jest.doMock('components/Header');

describe('Component - NewbieSelect', () => {
  it('renders correctly', async () => {
    const component = create(
      <MockedProvider mocks={newbieSelectMock} addTypename={false} resolvers={{}}>
        <AuthProvider value={mockedBuddyContext()}>
          <MenuProvider>
            <NewbieSelect />
          </MenuProvider>
        </AuthProvider>
      </MockedProvider>
    );

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await waitForExpect(() => {
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });
});
