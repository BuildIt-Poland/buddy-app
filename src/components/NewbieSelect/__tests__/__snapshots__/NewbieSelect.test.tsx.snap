// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - NewbieSelect renders correctly 1`] = `
Array [
  <Header
    loading={true}
    onButtonClick={[Function]}
    type="menu"
  />,
  <PageContainer
    data-testid="newbie-select-page"
  >
    <Box
      className="makeStyles-title-1"
      component="section"
    >
      <Typography
        component="h1"
        variant="h2"
      >
        Your New Joiners
      </Typography>
      <Typography
        color="textSecondary"
        component="p"
        variant="body2"
      >
        Don’t forget to complete all tasks
      </Typography>
    </Box>
    <PlusButton
      disabled={true}
    />
  </PageContainer>,
]
`;

exports[`Component - NewbieSelect renders correctly 2`] = `
Array [
  <Header
    loading={false}
    onButtonClick={[Function]}
    type="menu"
  />,
  <PageContainer
    data-testid="newbie-select-page"
  >
    <Box
      className="makeStyles-title-1"
      component="section"
    >
      <Typography
        component="h1"
        variant="h2"
      >
        Your New Joiners
      </Typography>
      <Typography
        color="textSecondary"
        component="p"
        variant="body2"
      >
        Don’t forget to complete all tasks
      </Typography>
    </Box>
    <PlusButton
      disabled={true}
    />
  </PageContainer>,
]
`;
